docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,833 @@
+Concurrency is a double-edged sword: while it could make Cesium blazing-fast, it introduces
+additional complexity through parallelism – when many subprocesses run at the same time,
+they create conflicts, or race conditions, when they attempt to access common resources.
+As Cesium looks to become production-ready, it is crucial that Cesium must support
+Read, Write, and Delete (through a tombstone & garbage collection system) operations
+at high speeds without them conflicting each other. One can imagine the possible
+consequences for unhandled race conditions in extreme scenarios such as launch control systems:
+it would not be desirable.
+
+# 4- Analysis
+
+## 4.0 Primer on the structure of Cesium
+
+As a primer for analyzing race conditions in Cesium, we will first analyze how Cesium
+is structured and how it stores data. While all the actual telemetry data is stored on
+the file system, the memory assists in storing information as to where on the file system
+a desired chunk of data may be found.
+
+### 4.0.1 Cesium's file storage structure
+
+All of Cesium's data is stored in the root directory specified by the user when Cesium
+is booted. Each channel stores its data in a subdirectory of the same name as the channel
+key: this subdirectory is the file system of that channel. Within this file system, a
+`meta.json` file is used to store the metadata of this channel, e.g. key, rate, etc.
+This file is not relevant for our purposes of analyzing race conditions.
+
+The channel's file system also includes the `index.domain` file used for domain
+index persist.
+All other files in the file system are used to store binary telemetry data – they are
+numerically labeled and more files get created as existing files get filled up – i.e.
+`1.domain`, `2.domain`, and so on.
+
+<div align="center">
+<img src="img/0019-240529-cesium-race-conditions/File-System-Structure.png" width="90%">
+<h6>File-System structure</h6>
+</div>
+
+### 4.0.2 Cesium's database structure
+
+At the highest level of abstraction, Cesium provides one unified interface to read, write,
+delete, and stream data from multiple channels.
+A Cesium DB comprises many channels, each one of which corresponding to a virtualDB
+or an unaryDB. A virtualDB contains no data, so it will not be a topic of our discussion.
+
+A unaryDB is a second level of abstraction in the database structure: it is composed of
+a domainDB an index, and the controller:
+the index is responsible for querying the timestamps corresponding to telemetry data,
+the controller is responsible for controlling writes to the database between different
+writers, and the domainDB is responsible for storing the actual telemetry data.
+
+A domainDB is the lowest level of abstraction: it provides means to write and read data from
+the file system, and maintains the domain index.
+
+<div align="center">
+<img src="img/0019-240529-cesium-race-conditions/Cesium-Structure.png" width="90%">
```

Comment:
In the graphic, I would remove the spaces between "domain DB" and "unary DB", since these correspond to variable names.